### ISSUE TITLE
test sample_distribution fails for wishart 

### ISSUE THREAD

**Issue by njtierney** [greta-dev/greta#414]
The test, [sample_distribution](https://github.com/greta-dev/greta/blob/master/tests/testthat/test_distributions.R#L697) for `wishart` fails as it cannot always be sampled from with MCMC.

Here is an expression of that, using `purrr::safely()` which returns either a result (NULL) or an error when there is an error. If it works all the time, we should expect that there would be nothing in the `error` row. Instead, we see that the `error` has entries.

``` r
library(greta)
#> 
#> Attaching package: 'greta'
#> The following objects are masked from 'package:stats':
#> 
#>     binomial, cov2cor, poisson
#> The following objects are masked from 'package:base':
#> 
#>     %*%, apply, backsolve, beta, chol2inv, colMeans, colSums, diag,
#>     eigen, forwardsolve, gamma, identity, rowMeans, rowSums, sweep,
#>     tapply
library(testthat)
library(purrr)
#> 
#> Attaching package: 'purrr'
#> The following object is masked from 'package:testthat':
#> 
#>     is_null

sample_distribution <- function(greta_array, n = 10,
                                lower = -Inf, upper = Inf,
                                warmup = 1) {
  m <- model(greta_array, precision = "double")
  draws <- mcmc(m, n_samples = n, warmup = warmup, verbose = FALSE)
  samples <- as.matrix(draws)
  vectorised <- length(lower) > 1 | length(upper) > 1
  
  if (vectorised) {
    above_lower <- sweep(samples, 2, lower, `>=`)
    below_upper <- sweep(samples, 2, upper, `<=`)
  } else {
    above_lower <- samples >= lower
    below_upper <- samples <= upper
  }
  
  expect_true(all(above_lower & below_upper))
}

safe_sample_distribution <- safely(sample_distribution)

sig <- rWishart(1, 4, diag(3))[, , 1]
wish_df <- function(df = 4) wishart(df, sig)
safe_replicate_samp_distr <- function(n_rep = 10, df = 4){
  replicate(n = 10, safe_sample_distribution(wish_df(df), warmup = 0))
}

library(tictoc)

tic()
(sample_wish_df_4 <- safe_replicate_samp_distr(4))
#>        [,1] [,2] [,3] [,4] [,5]   [,6] [,7]   [,8] [,9]   [,10]
#> result TRUE TRUE TRUE TRUE NULL   TRUE NULL   TRUE NULL   TRUE 
#> error  NULL NULL NULL NULL List,2 NULL List,2 NULL List,2 NULL
toc()
#> 20.052 sec elapsed

tic()
(sample_wish_df_10 <- safe_replicate_samp_distr(10))
#>        [,1]   [,2]   [,3] [,4]   [,5]   [,6] [,7] [,8] [,9]   [,10] 
#> result NULL   NULL   TRUE NULL   NULL   TRUE TRUE TRUE NULL   NULL  
#> error  List,2 List,2 NULL List,2 List,2 NULL NULL NULL List,2 List,2
toc()
#> 12.833 sec elapsed

tic()
(sample_wish_df_100 <- safe_replicate_samp_distr(100))
#>        [,1] [,2]   [,3] [,4] [,5]   [,6] [,7] [,8] [,9] [,10]
#> result TRUE NULL   TRUE TRUE NULL   TRUE TRUE TRUE TRUE TRUE 
#> error  NULL List,2 NULL NULL List,2 NULL NULL NULL NULL NULL
toc()
#> 20.786 sec elapsed
```

<sup>Created on 2021-05-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.5 (2021-03-31)
#>  os       macOS Big Sur 10.16         
#>  system   x86_64, darwin17.0          
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_AU.UTF-8                 
#>  ctype    en_AU.UTF-8                 
#>  tz       Australia/Perth             
#>  date     2021-05-11                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version    date       lib source            
#>  abind         1.4-5      2016-07-21 [1] CRAN (R 4.0.2)    
#>  backports     1.2.1      2020-12-09 [1] standard (@1.2.1) 
#>  base64enc     0.1-3      2015-07-28 [1] standard (@0.1-3) 
#>  cli           2.4.0      2021-04-05 [1] CRAN (R 4.0.2)    
#>  coda          0.19-4     2020-09-30 [1] CRAN (R 4.0.2)    
#>  codetools     0.2-18     2020-11-04 [1] CRAN (R 4.0.5)    
#>  crayon        1.4.1      2021-02-08 [1] CRAN (R 4.0.2)    
#>  digest        0.6.27     2020-10-24 [1] standard (@0.6.27)
#>  ellipsis      0.3.1      2020-05-15 [1] standard (@0.3.1) 
#>  evaluate      0.14       2019-05-28 [1] standard (@0.14)  
#>  fansi         0.4.2      2021-01-15 [1] CRAN (R 4.0.2)    
#>  fs            1.5.0      2020-07-31 [1] standard (@1.5.0) 
#>  future        1.21.0     2020-12-10 [1] CRAN (R 4.0.2)    
#>  globals       0.14.0     2020-11-22 [1] CRAN (R 4.0.2)    
#>  glue          1.4.2      2020-08-27 [1] standard (@1.4.2) 
#>  greta       * 0.3.1.9012 2021-05-06 [1] local             
#>  highr         0.8        2019-03-20 [1] standard (@0.8)   
#>  hms           1.0.0      2021-01-13 [1] CRAN (R 4.0.2)    
#>  htmltools     0.5.1.1    2021-01-22 [1] CRAN (R 4.0.2)    
#>  jsonlite      1.7.2      2020-12-09 [1] standard (@1.7.2) 
#>  knitr         1.31       2021-01-27 [1] CRAN (R 4.0.2)    
#>  lattice       0.20-41    2020-04-02 [1] CRAN (R 4.0.5)    
#>  lifecycle     1.0.0      2021-02-15 [1] CRAN (R 4.0.2)    
#>  listenv       0.8.0      2019-12-05 [1] CRAN (R 4.0.2)    
#>  magrittr      2.0.1      2020-11-17 [1] standard (@2.0.1) 
#>  Matrix        1.3-2      2021-01-06 [1] CRAN (R 4.0.5)    
#>  parallelly    1.24.0     2021-03-14 [1] CRAN (R 4.0.2)    
#>  pillar        1.6.0      2021-04-13 [1] CRAN (R 4.0.2)    
#>  pkgconfig     2.0.3      2019-09-22 [1] standard (@2.0.3) 
#>  prettyunits   1.1.1      2020-01-24 [1] standard (@1.1.1) 
#>  progress      1.2.2      2019-05-16 [1] CRAN (R 4.0.2)    
#>  purrr       * 0.3.4      2020-04-17 [1] standard (@0.3.4) 
#>  R6            2.5.0      2020-10-28 [1] standard (@2.5.0) 
#>  Rcpp          1.0.6      2021-01-15 [1] CRAN (R 4.0.2)    
#>  reprex        2.0.0      2021-04-02 [1] CRAN (R 4.0.2)    
#>  reticulate    1.19       2021-04-21 [1] CRAN (R 4.0.2)    
#>  rlang         0.4.10     2020-12-30 [1] CRAN (R 4.0.2)    
#>  rmarkdown     2.7        2021-02-19 [1] CRAN (R 4.0.2)    
#>  sessioninfo   1.1.1      2018-11-05 [1] standard (@1.1.1) 
#>  stringi       1.5.3      2020-09-09 [1] standard (@1.5.3) 
#>  stringr       1.4.0      2019-02-10 [1] standard (@1.4.0) 
#>  styler        1.4.1      2021-03-30 [1] CRAN (R 4.0.2)    
#>  tensorflow    2.4.0      2021-03-23 [1] CRAN (R 4.0.3)    
#>  testthat    * 3.0.2      2021-02-14 [1] CRAN (R 4.0.2)    
#>  tfruns        1.5.0      2021-02-26 [1] CRAN (R 4.0.2)    
#>  tibble        3.1.1      2021-04-18 [1] CRAN (R 4.0.3)    
#>  tictoc      * 1.0        2014-06-17 [1] CRAN (R 4.0.2)    
#>  utf8          1.2.1      2021-03-12 [1] CRAN (R 4.0.2)    
#>  vctrs         0.3.7      2021-03-29 [1] CRAN (R 4.0.3)    
#>  waldo         0.2.5      2021-03-08 [1] CRAN (R 4.0.2)    
#>  whisker       0.4        2019-08-28 [1] standard (@0.4)   
#>  withr         2.4.2      2021-04-18 [1] CRAN (R 4.0.3)    
#>  xfun          0.22       2021-03-11 [1] CRAN (R 4.0.2)    
#>  yaml          2.2.1      2020-02-01 [1] standard (@2.2.1) 
#> 
#> [1] /Library/Frameworks/R.framework/Versions/4.0/Resources/library
```

</details>

This issue could be addressed by addressing #413 (sampling initial values from priors by default) - what are your thoughts, @goldingn ?

**Comment by goldingn** [comment]
Try with just `sig <- diag(3)`?
Could be the covariance matrix there is quite far from diagonal, which would have a similar effect to the df

**Comment by njtierney** [comment]
``` r
library(greta)
#> 
#> Attaching package: 'greta'
#> The following objects are masked from 'package:stats':
#> 
#>     binomial, cov2cor, poisson
#> The following objects are masked from 'package:base':
#> 
#>     %*%, apply, backsolve, beta, chol2inv, colMeans, colSums, diag,
#>     eigen, forwardsolve, gamma, identity, rowMeans, rowSums, sweep,
#>     tapply
library(testthat)
library(purrr)
#> 
#> Attaching package: 'purrr'
#> The following object is masked from 'package:testthat':
#> 
#>     is_null

sample_distribution <- function(greta_array, n = 10,
                                lower = -Inf, upper = Inf,
                                warmup = 1) {
  m <- model(greta_array, precision = "double")
  draws <- mcmc(m, n_samples = n, warmup = warmup, verbose = FALSE)
  samples <- as.matrix(draws)
  vectorised <- length(lower) > 1 | length(upper) > 1
  
  if (vectorised) {
    above_lower <- sweep(samples, 2, lower, `>=`)
    below_upper <- sweep(samples, 2, upper, `<=`)
  } else {
    above_lower <- samples >= lower
    below_upper <- samples <= upper
  }
  
  expect_true(all(above_lower & below_upper))
}

safe_sample_distribution <- safely(sample_distribution)

# sig <- rWishart(1, 4, diag(3))[, , 1]
wish_df <- function(df = 4) wishart(df, Sig = diag(3))
safe_replicate_samp_distr <- function(n_rep = 10, df = 4){
  replicate(n = 10, safe_sample_distribution(wish_df(df), warmup = 0))
}

library(tictoc)

tic()
(sample_wish_df_4 <- safe_replicate_samp_distr(4))
#>        [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] 
#> result TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE NULL  
#> error  NULL NULL NULL NULL NULL NULL NULL NULL NULL List,2
toc()
#> 25.312 sec elapsed

tic()
(sample_wish_df_10 <- safe_replicate_samp_distr(10))
#>        [,1] [,2] [,3]   [,4]   [,5]   [,6] [,7]   [,8] [,9]   [,10]
#> result TRUE TRUE NULL   NULL   NULL   TRUE NULL   TRUE NULL   TRUE 
#> error  NULL NULL List,2 List,2 List,2 NULL List,2 NULL List,2 NULL
toc()
#> 15.036 sec elapsed

tic()
(sample_wish_df_100 <- safe_replicate_samp_distr(100))
#>        [,1] [,2]   [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
#> result TRUE NULL   TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE 
#> error  NULL List,2 NULL NULL NULL NULL NULL NULL NULL NULL
toc()
#> 26.155 sec elapsed
```

<sup>Created on 2021-05-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.5 (2021-03-31)
#>  os       macOS Big Sur 10.16         
#>  system   x86_64, darwin17.0          
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_AU.UTF-8                 
#>  ctype    en_AU.UTF-8                 
#>  tz       Australia/Perth             
#>  date     2021-05-11                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version    date       lib source            
#>  abind         1.4-5      2016-07-21 [1] CRAN (R 4.0.2)    
#>  backports     1.2.1      2020-12-09 [1] standard (@1.2.1) 
#>  base64enc     0.1-3      2015-07-28 [1] standard (@0.1-3) 
#>  cli           2.4.0      2021-04-05 [1] CRAN (R 4.0.2)    
#>  coda          0.19-4     2020-09-30 [1] CRAN (R 4.0.2)    
#>  codetools     0.2-18     2020-11-04 [1] CRAN (R 4.0.5)    
#>  crayon        1.4.1      2021-02-08 [1] CRAN (R 4.0.2)    
#>  digest        0.6.27     2020-10-24 [1] standard (@0.6.27)
#>  ellipsis      0.3.1      2020-05-15 [1] standard (@0.3.1) 
#>  evaluate      0.14       2019-05-28 [1] standard (@0.14)  
#>  fansi         0.4.2      2021-01-15 [1] CRAN (R 4.0.2)    
#>  fs            1.5.0      2020-07-31 [1] standard (@1.5.0) 
#>  future        1.21.0     2020-12-10 [1] CRAN (R 4.0.2)    
#>  globals       0.14.0     2020-11-22 [1] CRAN (R 4.0.2)    
#>  glue          1.4.2      2020-08-27 [1] standard (@1.4.2) 
#>  greta       * 0.3.1.9012 2021-05-06 [1] local             
#>  highr         0.8        2019-03-20 [1] standard (@0.8)   
#>  hms           1.0.0      2021-01-13 [1] CRAN (R 4.0.2)    
#>  htmltools     0.5.1.1    2021-01-22 [1] CRAN (R 4.0.2)    
#>  jsonlite      1.7.2      2020-12-09 [1] standard (@1.7.2) 
#>  knitr         1.31       2021-01-27 [1] CRAN (R 4.0.2)    
#>  lattice       0.20-41    2020-04-02 [1] CRAN (R 4.0.5)    
#>  lifecycle     1.0.0      2021-02-15 [1] CRAN (R 4.0.2)    
#>  listenv       0.8.0      2019-12-05 [1] CRAN (R 4.0.2)    
#>  magrittr      2.0.1      2020-11-17 [1] standard (@2.0.1) 
#>  Matrix        1.3-2      2021-01-06 [1] CRAN (R 4.0.5)    
#>  parallelly    1.24.0     2021-03-14 [1] CRAN (R 4.0.2)    
#>  pillar        1.6.0      2021-04-13 [1] CRAN (R 4.0.2)    
#>  pkgconfig     2.0.3      2019-09-22 [1] standard (@2.0.3) 
#>  prettyunits   1.1.1      2020-01-24 [1] standard (@1.1.1) 
#>  progress      1.2.2      2019-05-16 [1] CRAN (R 4.0.2)    
#>  purrr       * 0.3.4      2020-04-17 [1] standard (@0.3.4) 
#>  R6            2.5.0      2020-10-28 [1] standard (@2.5.0) 
#>  Rcpp          1.0.6      2021-01-15 [1] CRAN (R 4.0.2)    
#>  reprex        2.0.0      2021-04-02 [1] CRAN (R 4.0.2)    
#>  reticulate    1.19       2021-04-21 [1] CRAN (R 4.0.2)    
#>  rlang         0.4.10     2020-12-30 [1] CRAN (R 4.0.2)    
#>  rmarkdown     2.7        2021-02-19 [1] CRAN (R 4.0.2)    
#>  sessioninfo   1.1.1      2018-11-05 [1] standard (@1.1.1) 
#>  stringi       1.5.3      2020-09-09 [1] standard (@1.5.3) 
#>  stringr       1.4.0      2019-02-10 [1] standard (@1.4.0) 
#>  styler        1.4.1      2021-03-30 [1] CRAN (R 4.0.2)    
#>  tensorflow    2.4.0      2021-03-23 [1] CRAN (R 4.0.3)    
#>  testthat    * 3.0.2      2021-02-14 [1] CRAN (R 4.0.2)    
#>  tfruns        1.5.0      2021-02-26 [1] CRAN (R 4.0.2)    
#>  tibble        3.1.1      2021-04-18 [1] CRAN (R 4.0.3)    
#>  tictoc      * 1.0        2014-06-17 [1] CRAN (R 4.0.2)    
#>  utf8          1.2.1      2021-03-12 [1] CRAN (R 4.0.2)    
#>  vctrs         0.3.7      2021-03-29 [1] CRAN (R 4.0.3)    
#>  waldo         0.2.5      2021-03-08 [1] CRAN (R 4.0.2)    
#>  whisker       0.4        2019-08-28 [1] standard (@0.4)   
#>  withr         2.4.2      2021-04-18 [1] CRAN (R 4.0.3)    
#>  xfun          0.22       2021-03-11 [1] CRAN (R 4.0.2)    
#>  yaml          2.2.1      2020-02-01 [1] standard (@2.2.1) 
#> 
#> [1] /Library/Frameworks/R.framework/Versions/4.0/Resources/library
```

</details>

Unfortunately it is a no go

**Comment by njtierney** [comment]
OK added that into the test now, let's see how it goes.

**Comment by njtierney** [comment]
resolved in #412
